esrally/track/loader.py

```diff
@@ -1104,12 +1114,26 @@ def _create_corpora(self, corpora_specs, indices):
                         target_idx = None
                         target_type = None
                     else:
+                        target_ds = self._r(doc_spec, "target-data-stream", mandatory=False,
+                                             error_ctx=docs)
+
                         # we need an index if no meta-data are present.
-                        target_idx = self._r(doc_spec, "target-index", mandatory=corpus_target_idx is None,
-                                             default_value=corpus_target_idx, error_ctx=docs)
+                        target_idx = self._r(doc_spec, "target-index",
+                                             mandatory=corpus_target_idx is None and corpus_target_ds is None
+                                                       and target_ds is None,
+                                             error_ctx=docs)
                         target_type = self._r(doc_spec, "target-type", mandatory=False,
                                               default_value=corpus_target_type, error_ctx=docs)
 
+                        # here we choose to use either an index or data streams. If either are explicitly specified
+                        # (index takes precedence) this is preferred over any defaults. Index then takes precedence.
```

Comment:
I think this is a fair restriction if we consider it to be acceptable. It will simplify handling internally and more importantly be easier to reason about and document. Right now, we would have to describe precedence orders which i personally am not a fan of. +1 on either or.

@dliappis any opinion here?